apps/aecore/src/aec_chain_state.erl

```diff
@@ -335,6 +415,65 @@ internal_insert(Node, Block) ->
         {ok, Old} -> internal_error({same_key_different_content, Node, Old})
     end.
 
+
+assert_not_illegal_fork_or_orphan(Node, sync, State) ->
+    case node_is_genesis(Node, State) of
+        true -> ok;
+        false ->
+            PrevHash = prev_hash(Node),
+            case hash_is_connected_to_genesis(PrevHash) of
+                true  -> ok;
+                false -> internal_error({illegal_orphan, sync, hash(Node)})
+            end
+    end;
+assert_not_illegal_fork_or_orphan(Node, undefined, State) ->
+    case node_is_genesis(Node, State) of
+        true -> ok;
+        false ->
+            assert_height_delta(Node, State),
+            assert_connection_to_chain(Node, node_type(Node))
+    end.
+
+assert_connection_to_chain(Node, key) ->
+    case hash_is_connected_to_genesis(prev_hash(Node)) of
+        true  -> ok;
+        false -> internal_error({illegal_orphan, hash(Node)})
+    end;
+assert_connection_to_chain(Node, micro) ->
```

Comment:
this is good spam prevention, but in regards of PoF discussion - it will not prevent incomplete microforks. Just FYI, @hanssv @zp-sd